Build failure: zbar

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

Build zbar on aarch64-darwin (or [x86_64-darwin which also fails on Hydra](https://hydra.nixos.org/build/310232765))

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/310232763

### Relevant log output

```console
> ======================================================================
> FAIL: runTest (__main__.BuiltinTestCase.runTest)
> /private/tmp/nix-build-zbar-0.23.93.drv-0/source/examples/ean-13.png
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/private/tmp/nix-build-zbar-0.23.93.drv-0/source/test/barcodetest.py", line 171, in runTest
>     actual = run_zbarimg((expect.get('href'),))
>   File "/private/tmp/nix-build-zbar-0.23.93.drv-0/source/test/barcodetest.py", line 139, in run_zbarimg
>     assert rc in (0, 4), \
>            ^^^^^^^^^^^^
> AssertionError: zbarimg returned error status (-11):
>
>
>
> ----------------------------------------------------------------------
> Ran 1 test in 37.493s
>
> FAILED (failures=1)
> make[3]: *** [Makefile:2531: check-images-py] Error 1
> make[3]: Leaving directory '/private/tmp/nix-build-zbar-0.23.93.drv-0/source'
> make[2]: *** [Makefile:2217: check-am] Error 2
> make[2]: Leaving directory '/private/tmp/nix-build-zbar-0.23.93.drv-0/source'
> make[1]: *** [Makefile:1911: check-recursive] Error 1
> make[1]: Leaving directory '/private/tmp/nix-build-zbar-0.23.93.drv-0/source'
> make: *** [Makefile:2219: check] Error 2
```

### Additional context

https://github.com/mchehab/zbar/blob/a549566ea11eb03622bd4458a1728ffe3f589163/test/barcodetest.py#L119C5-L140
It seems like the test cases call `zbarimg`, which is failing

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.4.0, macOS 15.4`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.28.4`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/81d6lv96pws78iijmmsra5ql8wxgb1cb-nixpkgs/nixpkgs`

### Notify maintainers

@7c6f434c 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/tools/graphics/zbar/darwin-segfault-optimized-pointer-assignment.patch
+++ b/pkgs/tools/graphics/zbar/darwin-segfault-optimized-pointer-assignment.patch
@@ -1,0 +1,47 @@
+From 3fa414aa82375648635281924904557cbe4d2d83 Mon Sep 17 00:00:00 2001
+From: sasdf <asdf79852@gmail.com>
+Date: Fri, 18 Oct 2024 00:22:36 +0800
+Subject: [PATCH] Fix pointer wrap around undefined behavior
+
+---
+ zbar/img_scanner.c | 13 +++++++------
+ 1 file changed, 7 insertions(+), 6 deletions(-)
+
+diff --git a/zbar/img_scanner.c b/zbar/img_scanner.c
+index d1bf8a3c..58436f16 100644
+--- a/zbar/img_scanner.c
++++ b/zbar/img_scanner.c
+@@ -33,6 +33,7 @@
+ #endif
+ 
+ #include <assert.h>
++#include <stddef.h>
+ #include <stdlib.h> /* malloc, free */
+ #include <string.h> /* memcmp, memset, memcpy */
+ 
+@@ -50,7 +51,7 @@
+ #include "svg.h"
+ 
+ #if 1
+-#define ASSERT_POS assert(p == data + x + y * (intptr_t)w)
++#define ASSERT_POS assert(p == data + x + y * (ptrdiff_t)w)
+ #else
+ #define ASSERT_POS
+ #endif
+@@ -858,11 +859,11 @@ static void zbar_send_code_via_dbus(zbar_image_scanner_t *iscn,
+ }
+ #endif
+ 
+-#define movedelta(dx, dy)                \
+-    do {                                 \
+-	x += (dx);                       \
+-	y += (dy);                       \
+-	p += (dx) + ((uintptr_t)(dy)*w); \
++#define movedelta(dx, dy)                  \
++    do {                                   \
++        x += (dx);                         \
++        y += (dy);                         \
++        p += (dx) + ((dy)*(ptrdiff_t)(w)); \
+     } while (0);
+ 
+ static void *_zbar_scan_image(zbar_image_scanner_t *iscn, zbar_image_t *img)

--- a/pkgs/tools/graphics/zbar/default.nix
+++ b/pkgs/tools/graphics/zbar/default.nix
@@ -62,6 +62,11 @@ stdenv.mkDerivation rec {
       url = "https://github.com/mchehab/zbar/commit/a549566ea11eb03622bd4458a1728ffe3f589163.patch";
       hash = "sha256-NY3bAElwNvGP9IR6JxUf62vbjx3hONrqu9pMSqaZcLY=";
     })
+    # PR from fork not yet merged into upstream
+    # See PR: https://github.com/mchehab/zbar/pull/299
+    # Remove this patch if the PR is merged or if the issue is solved another way.
+    # See https://github.com/NixOS/nixpkgs/issues/456461 for discussion of the root issue
+    ./darwin-segfault-optimized-pointer-assignment.patch
   ];
 
   nativeBuildInputs = [

